File: LeetCodeCPlusPlus/py39.cpp
// 组合总和等于target的所有情况，可重复使用
#include <vector>
using namespace std;
class Solution {
    vector<vector<int>> result;
    vector<int> tmp;
private: 
    void helper(vector<int>& candidates, int target, int index) {
        if (target<0 || index>=candidates.size() || target!=0 && target<candidates[index])
            return;
        if (target == 0) {
            result.push_back(tmp);
        } else {
            helper(candidates, target, index+1);
            tmp.push_back(candidates[index]);
            helper(candidates, target-candidates[index], index);
            tmp.pop_back();
        }
    
}
public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        sort(candidates.begin(),candidates.end());
        helper(candidates, target, 0);
        return result;
    }
};

File: LeetCodePython/py39.py
# 组合总数， 和等于target的所有给定数字的组合方式，可重复。
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        result = []
        n = len(candidates)
        candidates.sort()
        def helper(index, total, current):
            if index>=n or total+candidates[index]>target:
                return 
            elif total + candidates[index] == target:
                current.append(candidates[index])
                result.append(list(current))
                current.pop()
                return
            else:
                helper(index+1, total ,current)
                current.append(candidates[index])
                helper(index, total+candidates[index], current)
                current.pop()
        helper(0,0,[])
        return result